server/core/src/storage.ts

```diff
@@ -299,73 +342,140 @@ class TServerStorage implements ServerStorage {
-    const isAttached = this.hierarchy.isDerived(tx.objectClass, core.class.AttachedDoc)
-    const txes = await this.findAll<TxCUD<Doc>>(
-      ctx,
-      isAttached ? core.class.TxCollectionCUD : core.class.TxCUD,
-      isAttached
-        ? { 'tx.objectId': tx.objectId as Ref<AttachedDoc> }
-        : {
-            objectId: tx.objectId
+  private async buildRemovedDoc (ctx: MeasureContext, rawTxes: Tx[], findAll: ServerStorage['findAll']): Promise<Doc[]> {
+    const removeObjectIds: Ref<Doc>[] = []
+    const removeAttachObjectIds: Ref<AttachedDoc>[] = []
+
+    for (const tx of rawTxes) {
+      const actualTx = TxProcessor.extractTx(tx)
+      if (!this.hierarchy.isDerived(actualTx._class, core.class.TxRemoveDoc)) {
+        continue
+      }
+      const rtx = actualTx as TxRemoveDoc<Doc>
+      const isAttached = this.hierarchy.isDerived(rtx.objectClass, core.class.AttachedDoc)
+      if (isAttached) {
+        removeAttachObjectIds.push(rtx.objectId as Ref<AttachedDoc>)
+      } else {
+        removeObjectIds.push(rtx.objectId)
+      }
+    }
+
+    const txes =
+      removeObjectIds.length > 0
+        ? await findAll<TxCUD<Doc>>(
+          ctx,
+          core.class.TxCUD,
+          {
+            objectId: { $in: removeObjectIds }
           },
-      { sort: { modifiedOn: 1 } }
-    )
-    const createTx = isAttached
-      ? txes.find((tx) => (tx as TxCollectionCUD<Doc, AttachedDoc>).tx._class === core.class.TxCreateDoc)
-      : txes.find((tx) => tx._class === core.class.TxCreateDoc)
-    if (createTx === undefined) return
-    let doc = TxProcessor.createDoc2Doc(createTx as TxCreateDoc<Doc>)
-    for (let tx of txes) {
-      tx = TxProcessor.extractTx(tx) as TxCUD<Doc>
-      if (tx._class === core.class.TxUpdateDoc) {
-        doc = TxProcessor.updateDoc2Doc(doc, tx as TxUpdateDoc<Doc>)
-      } else if (tx._class === core.class.TxMixin) {
-        const mixinTx = tx as TxMixin<Doc, Doc>
-        doc = TxProcessor.updateMixin4Doc(doc, mixinTx)
+          { sort: { modifiedOn: 1 } }
+        )
+        : []
+    const result: Doc[] = []
+
+    const txesAttach =
+      removeAttachObjectIds.length > 0
+        ? await findAll<TxCollectionCUD<Doc, AttachedDoc>>(
+          ctx,
+          core.class.TxCollectionCUD,
+          { 'tx.objectId': { $in: removeAttachObjectIds } },
+          { sort: { modifiedOn: 1 } }
+        )
+        : []
+
+    for (const tx of rawTxes) {
+      const actualTx = TxProcessor.extractTx(tx)
+      if (!this.hierarchy.isDerived(actualTx._class, core.class.TxRemoveDoc)) {
+        continue
+      }
+      const rtx = actualTx as TxRemoveDoc<Doc>
+      const isAttached = this.hierarchy.isDerived(rtx.objectClass, core.class.AttachedDoc)
+
+      const objTxex = isAttached
+        ? txesAttach.filter((tx) => tx.tx.objectId === rtx.objectId)
+        : txes.filter((it) => it.objectId === rtx.objectId)
+
+      const createTx = isAttached
```

Comment:
We have `TxProcessor.buildDoc2Doc` for next 15 lines

Reply:
Thanks.